src/backend/distributed/worker/worker_create_or_replace.c

```diff
@@ -0,0 +1,179 @@
+/*-------------------------------------------------------------------------
+ *
+ * worker_create_if_not_exist.c
+ *   TODO rename file and document, was named after old function
+ *
+ * Copyright (c) 2019, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "catalog/dependency.h"
+#include "catalog/pg_type.h"
+#include "fmgr.h"
+#include "nodes/makefuncs.h"
+#include "nodes/nodes.h"
+#include "parser/parse_type.h"
+#include "tcop/dest.h"
+#include "tcop/utility.h"
+#include "utils/builtins.h"
+#include "utils/regproc.h"
+
+#include "distributed/commands.h"
+#include "distributed/commands/utility_hook.h"
+#include "distributed/deparser.h"
+#include "distributed/metadata/distobject.h"
+#include "distributed/worker_protocol.h"
+
+static Node * CreateStmtByObjectAddress(const ObjectAddress *address);
+static DropStmt * drop_stmt_from_object_create(Node *createStmt);
+
+
+PG_FUNCTION_INFO_V1(worker_create_or_replace);
+
+
+/*
+ * worker_create_or_replace(statement text)
+ *
+ * function is called, by the coordinator, with a CREATE statement for an object. This
+ * function implements the CREATE ... IF NOT EXISTS functionality for objects that do not
+ * have this functionality or where their implementation is not sufficient.
+ *
+ * Besides checking if an object of said name exists it tries to compare the object to be
+ * created with the one in the local catalog. If there is a difference the on in the local
+ * catalog will be renamed after which the statement can be executed on this worker to
+ * create the object.
+ *
+ * Renaming has two purposes
+ *  - free the identifier for creation
+ *  - non destructive if there is data store that would be destroyed if the object was
+ *    used in a table on this node, eg. types. If the type would be dropped with a cascade
+ *    it would drop any column holding user data for this type.
+ */
+Datum
+worker_create_or_replace(PG_FUNCTION_ARGS)
+{
+	text *sqlStatementText = PG_GETARG_TEXT_P(0);
+	const char *sqlStatement = text_to_cstring(sqlStatementText);
+	const ObjectAddress *address = NULL;
+	Node *parseTree = ParseTreeNode(sqlStatement);
+
+	/*
+	 * since going to the drop statement might require some resolving we will do a check
+	 * if the type actually exists instead of adding the IF EXISTS keyword to the
+	 * statement.
+	 */
+	address = GetObjectAddressFromParseTree(parseTree, true);
+	if (ObjectExists(address))
+	{
+		Node *localCreateStmt = NULL;
+		const char *localSqlStatement = NULL;
+		DropStmt *dropStmtParseTree = NULL;
+
+		localCreateStmt = CreateStmtByObjectAddress(address);
+		localSqlStatement = DeparseTreeNode(localCreateStmt);
+		if (strcmp(sqlStatement, localSqlStatement) == 0)
+		{
+			/*
+			 * TODO string compare is a poor mans comparison, but calling equal on the
```

Comment:
Is it possible that we fail to detect this because of `QualifyTreeNode ` on the coordinator?

Reply:
`CreateStmtByObjectAddress` is supposed to fully qualify the statement.